smarts/core/coordinates.py

```diff
@@ -118,6 +120,144 @@ def __del__(self):
             del _shapely_points[self]
 
 
+def euclidean_distance(p1: Point, p2: Point) -> float:
+    """The distance taking measuring a direct line between p1 and p2."""
+    dx = p1[0] - p2[0]
+    dy = p1[1] - p2[1]
+    return math.sqrt(dx * dx + dy * dy)
+
+
+def position_at_offset(p1: Point, p2: Point, offset: float) -> Optional[Point]:
+    """A point between p1 and p2 given an offset less than the distance between p1 and p2."""
+    if is_close(offset, 0.0):  # for pathological cases with dist == 0 and offset == 0
+        return p1
+
+    dist = euclidean_distance(p1, p2)
+
+    if is_close(dist, offset):
+        return p2
+
+    if offset > dist:
+        return None
+
+    x = p1[0] + (p2[0] - p1[0]) * (offset / dist)
+    y = p1[1] + (p2[1] - p1[1]) * (offset / dist)
+    return Point(x, y)
+
+
+def offset_along_shape(point: Point, shape: List[Point]) -> Union[float, int]:
+    """An offset on a shape defined as a vector path determined by the closest location on the
+    path to the point.
+    """
+    if point not in shape:
+        return polygon_offset_with_minimum_distance_to_point(point, shape)
+    offset = 0
+    for i in range(len(shape) - 1):
+        if shape[i] == point:
+            break
+        offset += euclidean_distance(shape[i], shape[i + 1])
+    return offset
+
+
+def position_at_shape_offset(shape: List[Point], offset: float) -> Optional[Point]:
+    """A point defined as the offset into a shape defined as vector path."""
+    seen_length = 0
+    curr = shape[0]
+    for next_p in shape[1:]:
+        next_length = euclidean_distance(curr, next_p)
+        if seen_length + next_length > offset:
+            return position_at_offset(curr, next_p, offset - seen_length)
+        seen_length += next_length
+        curr = next_p
+    return shape[-1]
+
+
+def line_offset_with_minimum_distance_to_point(
+    point: Point,
+    line_start: Point,
+    line_end: Point,
+    perpendicular: bool = False,
+) -> Union[float, int]:
+    """Return the offset from line (line_start, line_end) where the distance to
+    point is minimal"""
+    p = point
+    p1 = line_start
+    p2 = line_end
+    d = euclidean_distance(p1, p2)
+    u = ((p[0] - p1[0]) * (p2[0] - p1[0])) + ((p[1] - p1[1]) * (p2[1] - p1[1]))
+    if d == 0.0 or u < 0.0 or u > d * d:
+        if perpendicular:
+            return -1
+        if u < 0.0:
+            return 0.0
+        return d
```

Comment:
This can be simplified slightly.

```suggestion
    if u < 0.0:
        return 0.0
    if d == 0.0 or u > d * d:
        if perpendicular:
            return -1
        return d
```